src/WooCommerce.php

```diff
@@ -398,6 +398,22 @@ public static function wp_head() {
     }
   }
 
+  /*
+   * Adds preload tag for main product image to improve largest contentful paint.
+   *
+   * @implements wp_head
+   */
+  public static function preloadMainProductImage() {
+    global $post;
+
+    if ($post && $post->post_type === 'product') {
+      $image = wp_get_attachment_image_src(get_post_thumbnail_id($post->ID), 'woocommerce_single');
+      if ($image) {
+        echo '<link rel="preload" href="' . $image[0] . '" as="image">';
+      }
+    }
+  }
```

Comment:
You can get rid of the conditional nesting doing an early return.

```suggestion
  public static function preloadMainProductImage() {
    global $post;
    
    if (!$post || !$post->post_type === 'product') {
      return;
    }

    $image = wp_get_attachment_image_src(get_post_thumbnail_id($post->ID), 'woocommerce_single');
    if ($image) {
      echo '<link rel="preload" href="' . $image[0] . '" as="image">';
    }
  }
```